### PR TITLE
Respect robots

### PR DESCRIPTION
**Pull request by brendt**
Addresses #114

### CODE REVIEW

File: src/Crawler.php [hunk 1]
```diff
@@ -61,6 +68,9 @@ class Crawler
     /** @var Browsershot */
     protected $browsershot = null;
 
+    /** @var \Spatie\Robots\RobotsTxt */
+    private $robotsTxt = null;
```
Comment (freekmurze) [inline src/Crawler.php]: make this `protected`

File: src/Crawler.php [hunk 2]
```diff
@@ -145,6 +155,18 @@ public function setMaximumDepth(int $maximumDepth)
         return $this;
     }
 
+    /**
+     * @param bool $ignoreRobots
+     *
+     * @return $this
+     */
+    public function ignoreRobots(bool $ignoreRobots = true)
```
Comment (freekmurze) [inline src/Crawler.php]: Do not accept a boolean here, let's add a dedicated `respectRobots` method.

File: src/Crawler.php [hunk 3]
```diff
@@ -301,9 +338,14 @@ protected function convertBodyToString(StreamInterface $bodyStream, $readMaximum
...
 
+    protected function createRobotsTxt(UriInterface $uri): RobotsTxt
+    {
+        return RobotsTxt::create($uri->withPath('/robots.txt'));
+    }
+
     /**
      * @param ResponseInterface|null $response
-     * @param CrawlUrl $crawlUrl
+     * @param CrawlUrl               $crawlUrl
```
Comment (freekmurze) [inline src/Crawler.php]: I don't like this indentation, but I guess StyleCI will take care of this.

File: src/Crawler.php [hunk 4]
```diff
@@ -393,6 +435,10 @@ protected function addAllLinksToCrawlQueue(string $html, UriInterface $foundOnUr
 
     protected function shouldCrawl(Node $node): bool
     {
+        if (! $this->ignoreRobots && ! $this->robotsTxt->allows($node->getValue())) {
```
Comment (freekmurze) [inline src/Crawler.php]: Remove `&&` by using an extra `if`

File: src/Crawler.php [hunk 5]
```diff
@@ -505,4 +551,22 @@ protected function maximumCrawlCountReached(): bool
...
         return $this->crawledUrlCount >= $this->maximumCrawlCount;
     }
+
+    protected function mayIndex(RobotsHeaders  $robotsHeaders, RobotsMeta $robotsMeta): bool
+    {
+        if ($this->ignoreRobots) {
+            return true;
+        }
+
+        return $robotsHeaders->mayIndex() && $robotsMeta->mayIndex();
```
Comment (freekmurze) [inline src/Crawler.php]: Avoid `&&` but using an extra `if`

File: src/Crawler.php [hunk 6]
```diff
@@ -238,6 +260,8 @@ public function startCrawling($baseUrl)
 
         $crawlUrl = CrawlUrl::create($this->baseUrl);
 
+        $this->robotsTxt = $this->createRobotsTxt($crawlUrl->url);
```
Comment (freekmurze) [inline src/Crawler.php]: Will this work when the given url is not the root of a domain?
eg `https://example.com/page/subpage`

File: README.md [hunk 7]
```diff
@@ -155,6 +155,16 @@ This package comes with three `CrawlProfiles` out of the box:
 - `CrawlInternalUrls`: this profile will only crawl the internal urls on the pages of a host.
 - `CrawlSubdomainUrls`: this profile will only crawl the internal urls and its subdomains on the pages of a host.
 
+### Ignoring robots.txt and robots meta
+
+By default, the crawler will respect robots data. It is possible to disable these checks like so:
```
Comment (freekmurze) [inline README.md]: Add some extra info here on what a robots.txt and header is? Also mention that the actual check is being done by our robots-txt package.

File: src/Crawler.php (hunk 6 above)
Comment (brendt) [reply src/Crawler.php]: Yes, the URL is used in `createRobotsTxt` like so:

```
$uri->withPath('/robots.txt')
```

By the robots spec, the `robots.txt` must always be in the root.

File: src/Crawler.php (hunk 3 above)
Comment (brendt) [reply src/Crawler.php]: I've disabled it in my auto formatter :)

File: src/Crawler.php [hunk 8]
```diff
@@ -52,6 +55,9 @@ class Crawler
     /** @var int|null */
     protected $maximumDepth = null;
 
+    /** @var bool */
+    protected $ignoreRobots = false;
```
Comment (freekmurze) [inline src/Crawler.php]: I think the code will be clearer if we use something positive here. Let's rename it to `respectRobots` and set it to `true`.

File: README.md (hunk 7 above)
Comment (brendt) [reply README.md]: Also done.

**Comment by brendt** [comment]
@freekmurze I fixed the failing tests and think it's ready to merge now. One question though: how will we go about testing this for a while without tagging a release?

**Pull request merged**